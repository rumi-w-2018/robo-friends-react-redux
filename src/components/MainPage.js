import React, { Component } from 'react';
import Cards from './Cards';
import SearchBox from './SearchBox';
import Scroll from './Scroll';
import ErrorBoundary from './ErrorBoundary';
import Header from './Header';
import './MainPage.css';

class MainPage extends Component {
  componentDidMount() {
    this.props.getRobots();
  }

  filterRobots = () => {
    const { searchKey, robots } = this.props; // Destructuring: this.props.robots becomes robots
    return robots.data.filter(robot => {
      return robot.name.toLowerCase().includes(searchKey);
    });
  };

  render() {
    const { robots } = this.props; // Destructuring: this.props.robots becomes robots

    if (robots.isPending) {
      return <h1 className="white pa3">Please wait...</h1>;
    } else if (!robots || robots.isError || robots.data.length === 0) {
      return <h1 className="white pa3">No robots were retrieved.</h1>;
    }

    return (
      <div className="animation-effect pv5 ph2 ma0 tc">
        <Header />
        <SearchBox onSearchChange={this.onSearchChange} />
        <hr className="custom" />
        <Scroll>
          <ErrorBoundary>
            <Cards robots={this.filterRobots()} />
          </ErrorBoundary>
        </Scroll>
      </div>
    );
  }
}

export default MainPage;
