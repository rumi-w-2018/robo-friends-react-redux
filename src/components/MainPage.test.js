import { shallow } from 'enzyme';
import React from 'react';
import MainPage from './MainPage';

let wrapper;
beforeEach( () => {
    const mockProps = {
        getRobots: jest.fn(),
        robots: {
            isPending: false,
            isError: false,
            data: [] },
        searchKey: ''
    };
    wrapper = shallow(<MainPage { ...mockProps } />);
});

it('renders MainPage without crashing', () => {
    expect(wrapper).toMatchSnapshot();
});

it('can be pending', () => {
    const mockProps = {
        getRobots: jest.fn(),
        robots: {
            isPending: true,
            isError: false,
            data: [] },
        searchKey: 'eee'
    };

    wrapper = shallow(<MainPage { ...mockProps } />);
    expect(wrapper.instance().filterRobots()).toEqual([]);
});

it('filters robots correctly', () => {
    const mockProps = {
        getRobots: jest.fn(),
        robots: {
            isPending: false,
            isError: false,
            data: [] },
        searchKey: 'eee'
    };

    wrapper = shallow(<MainPage { ...mockProps } />);
    expect(wrapper.instance().filterRobots()).toEqual([]);
});

it('filters robots correctly', () => {
    const mockProps = {
        getRobots: jest.fn(),
        robots: {
            isPending: false,
            isError: false,
            data: [{
                id: 1,
                name: 'Ervin Howell',
                username: 'Shanna',
                email: 'Shanna@melissa.tv'
            }] 
        },
        searchKey: 'ervin'
    };

    const filteredRobot = [{
        id: 1,
        name: 'Ervin Howell',
        username: 'Shanna',
        email: 'Shanna@melissa.tv'
    }];

    wrapper = shallow(<MainPage { ...mockProps } />);
    expect(wrapper.instance().filterRobots()).toEqual(filteredRobot);

});

it('filters robots correctly2', () => {
    const mockProps = {
        getRobots: jest.fn(),
        robots: {
            isPending: false,
            isError: false,
            data: [{
                id: 1,
                name: 'Ervin Howell',
                username: 'Shanna',
                email: 'Shanna@melissa.tv'
            }] 
        },
        searchKey: 'good'
    };

    const filteredRobot = [{
        id: 1,
        name: 'Ervin Howell',
        username: 'Shanna',
        email: 'Shanna@melissa.tv'
    }];

    wrapper = shallow(<MainPage { ...mockProps } />);
    expect(wrapper.instance().filterRobots()).toEqual([]);

});

