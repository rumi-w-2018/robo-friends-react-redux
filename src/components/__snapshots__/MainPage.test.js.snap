// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders MainPage without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainPage
    getRobots={[Function]}
    robots={
        Object {
            "data": Array [],
            "isError": false,
            "isPending": false,
          }
    }
    searchKey=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "No robots were retrieved.",
      "className": "white pa3",
    },
    "ref": null,
    "rendered": "No robots were retrieved.",
    "type": "h1",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "No robots were retrieved.",
        "className": "white pa3",
      },
      "ref": null,
      "rendered": "No robots were retrieved.",
      "type": "h1",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
